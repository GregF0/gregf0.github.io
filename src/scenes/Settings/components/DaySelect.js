import React, { useEffect, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import styled, { css } from 'styled-components'
import Box from '@material-ui/core/Box'
import MatFormControl from '@material-ui/core/FormControl'
import Select from '@material-ui/core/Select'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import { startSetFirstDayOfTheWeek } from '../../../data/settings/actions'
import { useTheme } from '@material-ui/core'

export const DaySelect = () => {
  const [day, setDay] = useState('')

  const firstDayOfTheWeek = useSelector(
    (state) => state.settings.firstDayOfTheWeek
  )

  const darkMode = useSelector((state) => +state.settings.darkMode)

  useEffect(() => {
    if (firstDayOfTheWeek !== null && firstDayOfTheWeek !== undefined) {
      setDay(firstDayOfTheWeek)
    }
  }, [firstDayOfTheWeek])

  const dispatch = useDispatch()

  const handleChange = (event) => {
    const day = event.target.value

    setDay(day)
    dispatch(startSetFirstDayOfTheWeek(day))
  }

  const theme = useTheme()

  return (
    <Box my={2}>
      <FormControl variant="outlined" dark={darkMode} theme={theme}>
        <InputLabel id="day-select">Week starts on</InputLabel>
        <Select
          labelId="day-select"
          value={day}
          onChange={handleChange}
          label="Week starts on"
        >
          <MenuItem value="Monday">Monday</MenuItem>
          <MenuItem value="Sunday">Sunday</MenuItem>
        </Select>
      </FormControl>
    </Box>
  )
}

const FormControl = styled(MatFormControl)`
  width: 100%;

  ${({ dark, theme }) =>
    dark &&
    css`
      .MuiFormLabel-root.Mui-focused {
        color: ${theme.palette.primary.light};
      }

      .MuiOutlinedInput-root.Mui-focused .MuiOutlinedInput-notchedOutline {
        border-color: ${theme.palette.primary.light};
      }
    `}
`
