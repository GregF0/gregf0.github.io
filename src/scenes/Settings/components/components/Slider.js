import React, { useState, useEffect } from 'react'
import styled, { css } from 'styled-components'
import Typography from '@material-ui/core/Typography'
import MatSlider from '@material-ui/core/Slider'
import Box from '@material-ui/core/Box'
import { useDispatch, useSelector } from 'react-redux'
import { useTheme } from '@material-ui/core'

export const Slider = ({
  name,
  step,
  marks,
  min,
  max,
  unit,
  action,
  value,
}) => {
  const [sliderValue, setSliderValue] = useState(value)

  useEffect(() => {
    setSliderValue(value)
  }, [value])

  const dispatch = useDispatch()

  const handleChangeCommitted = (event, newValue) => {
    if (newValue !== value) {
      dispatch(action(newValue))
    }
  }

  const handleChange = (event, newValue) => {
    setSliderValue(newValue)
  }

  const valueText = (value) => {
    return `${value} ${unit}`
  }

  const theme = useTheme()
  const darkMode = useSelector((state) => +state.settings.darkMode)
  const darkModeCached = +JSON.parse(localStorage.getItem('darkMode'))

  return (
    <Box mt={2}>
      <Typography id={`${name}-slider`} gutterBottom>
        {name}
      </Typography>
      <SettingSlider
        getAriaValueText={valueText}
        aria-labelledby={`${name}-slider`}
        valueLabelDisplay="auto"
        step={step}
        marks={marks}
        min={min}
        max={max}
        onChangeCommitted={handleChangeCommitted}
        onChange={handleChange}
        value={sliderValue}
        theme={theme}
        dark={darkMode || darkModeCached}
        color="primary"
      />
    </Box>
  )
}

export const SettingSlider = styled(MatSlider)`
  ${({ dark, theme }) =>
    dark &&
    css`
      color: ${theme.palette.primary.light};
    `}
`
