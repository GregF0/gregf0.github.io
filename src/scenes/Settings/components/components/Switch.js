import React, { useState, useEffect } from 'react'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import MatSwitch from '@material-ui/core/Switch'
import { useDispatch, useSelector } from 'react-redux'
import styled, { css } from 'styled-components'
import Box from '@material-ui/core/Box'
import { useTheme } from '@material-ui/core'

export const Switch = ({ name, action, checked }) => {
  const [state, setState] = useState(false)

  const handleChange = (event) => {
    const checked = event.target.checked
    setState(checked)
    dispatch(action(checked))
  }

  const dispatch = useDispatch()

  const theme = useTheme()
  const darkMode = useSelector((state) => +state.settings.darkMode)
  const darkModeCached = +JSON.parse(localStorage.getItem('darkMode'))

  useEffect(() => {
    if (checked !== null && checked !== undefined) {
      setState(checked)
    }
  }, [checked])

  return (
    <Box>
      <Label
        ml={0}
        control={
          <SettingSwitch
            checked={state}
            onChange={handleChange}
            name={name}
            theme={theme}
            color="primary"
            dark={darkMode || darkModeCached}
          />
        }
        label={name}
        labelPlacement="start"
      />
    </Box>
  )
}

export const Label = styled(FormControlLabel)`
  margin: 0;
`

export const SettingSwitch = styled(MatSwitch)`
  ${({ dark, theme }) =>
    dark &&
    css`
      .Mui-checked,
      .Mui-checked + .MuiSwitch-track {
        color: ${theme.palette.primary.light};
      }
    `}
`
