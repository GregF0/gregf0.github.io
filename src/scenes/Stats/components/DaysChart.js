import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Box from '@material-ui/core/Box'
import MatCard from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import IconButton from '@material-ui/core/IconButton'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import Typography from '@material-ui/core/Typography'
import MatCardContent from '@material-ui/core/CardContent'
import CardActions from '@material-ui/core/CardActions'
import Chip from '@material-ui/core/Chip'
import { Doughnut } from 'react-chartjs-2'
import dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import { daysOfWeek, colors } from '../data/constants'
import {
  generateChartTitle,
  generateChartTimeLabel,
  generateChartSessionsLabel,
} from '../data/chartFunctions'
import { filters, getFilterFn } from '../data/filters'

dayjs.extend(isSameOrAfter)
require('dayjs/locale/en-gb')
require('dayjs/locale/en')

export const DaysChart = () => {
  const sessions = useSelector((state) => state.sessions)
  const darkMode = useSelector((state) => +state.settings.darkMode)
  const darkModeCached = +localStorage.getItem('darkMode')

  const firstDayOfTheWeek = useSelector(
    (state) => state.settings.firstDayOfTheWeek
  )

  const [chartData, setChartData] = useState({
    labels: daysOfWeek,
    datasets: [
      {
        data: Array(7).fill(1),
        backgroundColor: [],
      },
    ],
  })

  const theme = useTheme()

  const [chartOptions, setChartOptions] = useState({
    maintainAspectRatio: false,
    legend: {
      labels: {
        fontColor: theme.palette.text.secondary,
        boxWidth: 20,
      },
    },
    tooltips: {
      callbacks: {
        title: generateChartTitle,
        label: generateChartTimeLabel,
      },
    },
  })

  const [calculatedData, setCalculatedData] = useState()
  const [dataType, setDataType] = useState('time')
  const [anchorEl, setAnchorEl] = useState(null)
  const [filter, setFilter] = useState(filters.find((f) => f.default))

  useEffect(() => {
    if (sessions && sessions.length && firstDayOfTheWeek) {
      let week = [...daysOfWeek]

      if (firstDayOfTheWeek === 'Monday') {
        dayjs.locale('en-gb')
      } else {
        dayjs.locale('en')
      }

      const filterFn = getFilterFn(filter)

      const data = Array(7)
        .fill()
        .map(() => ({ sessions: 0, time: 0 }))

      sessions.forEach(({ createdAt, duration }) => {
        if (filterFn(createdAt, dayjs)) {
          const dayIndex = dayjs(createdAt).day()

          data[dayIndex] = {
            sessions: data[dayIndex].sessions + 1,
            time: data[dayIndex].time + getSeconds(duration),
          }
        }
      })

      if (firstDayOfTheWeek === 'Monday') {
        const sunday = data[0]
        data.splice(0, 1)
        data.push(sunday)

        week.splice(0, 1)
        week.push('Sunday')
      }

      setCalculatedData(data)

      setChartData((prev) => {
        const newData = { ...prev }
        newData.datasets[0].data = data.map((d) => d[dataType])

        newData.labels = week
        return newData
      })
    }
  }, [sessions, firstDayOfTheWeek, filter, dataType])

  const onChipClicked = (typeSelected) => {
    if (typeSelected === dataType) return

    setDataType(typeSelected)

    setChartData((prev) => {
      const newData = { ...prev }
      newData.datasets[0].data = calculatedData.map(
        (data) => data[typeSelected]
      )
      return newData
    })

    setChartOptions((prev) => {
      const newOptions = { ...prev }

      newOptions.tooltips.callbacks.label =
        typeSelected === 'time'
          ? generateChartTimeLabel
          : generateChartSessionsLabel

      return newOptions
    })
  }

  useEffect(() => {
    setChartOptions((prev) => {
      const newData = { ...prev }
      newData.legend.labels.fontColor = theme.palette.text.secondary
      return newData
    })

    setChartData((prev) => {
      const newData = { ...prev }
      newData.datasets[0].backgroundColor = colors.map((c) =>
        darkModeCached ? c.darkMode : c.normal
      )
      return newData
    })
  }, [darkModeCached, theme.palette.text.secondary])

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const closeMenu = () => {
    setAnchorEl(null)
  }

  const handleSelect = (filter) => {
    setFilter(filter)
    closeMenu()
  }

  const getSeconds = ({ minutes, seconds }) => minutes * 60 + seconds

  return (
    <Card theme={theme} dark={darkMode}>
      <CardHeader
        title="Days of week"
        action={
          <IconButton aria-label="filtering options" onClick={openMenu}>
            <MoreVertIcon />
          </IconButton>
        }
      ></CardHeader>

      <Menu
        id="long-menu"
        anchorEl={anchorEl}
        open={!!anchorEl}
        onClose={closeMenu}
        margin="dense"
      >
        {filters.map((f, i) => (
          <MenuItem
            selected={filter.value === f.value}
            key={i}
            onClick={() => handleSelect(f)}
          >
            {f.name}
          </MenuItem>
        ))}
      </Menu>

      <CardContent>
        <Typography color="textSecondary">
          {filter.displayName || filter.name || ''}
        </Typography>

        <ChartContainer>
          <Doughnut data={chartData} options={chartOptions} />
        </ChartContainer>
      </CardContent>

      <CardActions>
        <div>
          <Box mr={1} display="inline-block">
            <Chip
              onClick={() => onChipClicked('time')}
              label="Time"
              color={dataType === 'time' ? 'primary' : 'default'}
              clickable
            />
          </Box>

          <Chip
            onClick={() => onChipClicked('sessions')}
            label="Sessions"
            color={dataType === 'sessions' ? 'primary' : 'default'}
            clickable
          />
        </div>
      </CardActions>
    </Card>
  )
}

const Card = styled(MatCard)`
  background: ${({ dark }) => (dark ? '#252525' : '#fff')};
`

const CardContent = styled(MatCardContent)`
  padding-top: 0;

  .MuiTypography-colorTextSecondary {
    font-weight: bold;
    font-size: 0.9rem;
  }
`

const ChartContainer = styled.div`
  height: 250px;
  max-width: 500px;
  margin: auto;
`
