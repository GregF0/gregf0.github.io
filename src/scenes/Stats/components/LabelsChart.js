import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import styled from 'styled-components'
import { useTheme } from '@material-ui/core'
import Box from '@material-ui/core/Box'
import MatCard from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import IconButton from '@material-ui/core/IconButton'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import Menu from '@material-ui/core/Menu'
import MenuItem from '@material-ui/core/MenuItem'
import Typography from '@material-ui/core/Typography'
import MatCardContent from '@material-ui/core/CardContent'
import CardActions from '@material-ui/core/CardActions'
import Chip from '@material-ui/core/Chip'
import * as materialColors from '@material-ui/core/colors'
import { Doughnut } from 'react-chartjs-2'
import dayjs from 'dayjs'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import { filters, getFilterFn } from '../data/filters'
import {
  generateChartTitle,
  generateChartTimeLabel,
  generateChartSessionsLabel,
} from '../data/chartFunctions'

dayjs.extend(isSameOrAfter)
require('dayjs/locale/en-gb')
require('dayjs/locale/en')

export const LabelsChart = () => {
  const labels = useSelector((state) => state.labels.data)
  const sessions = useSelector((state) => state.sessions)
  const darkMode = useSelector((state) => +state.settings.darkMode)
  const darkModeCached = +localStorage.getItem('darkMode')

  const firstDayOfTheWeek = useSelector(
    (state) => state.settings.firstDayOfTheWeek
  )

  const [chartData, setChartData] = useState({
    labels: [],
    datasets: [{ data: [], backgroundColor: [] }],
  })

  const theme = useTheme()

  const [chartOptions, setChartOptions] = useState({
    maintainAspectRatio: false,
    legend: {
      labels: {
        fontColor: theme.palette.text.secondary,
        boxWidth: 20,
      },
    },
    tooltips: {
      callbacks: {
        title: generateChartTitle,
        label: generateChartTimeLabel,
      },
    },
  })

  const [calculatedData, setCalculatedData] = useState()
  const [dataType, setDataType] = useState('time')
  const [anchorEl, setAnchorEl] = useState(null)
  const [filter, setFilter] = useState(filters.find((f) => f.default))

  useEffect(() => {
    setChartOptions((prev) => {
      const newData = { ...prev }
      newData.legend.labels.fontColor = theme.palette.text.secondary
      return newData
    })

    setChartData((prev) => {
      const newData = { ...prev }

      newData.datasets[0].backgroundColor = [
        ...newData.datasets[0].backgroundColor,
      ]

      return newData
    })
  }, [darkModeCached, theme.palette.text.secondary])

  useEffect(() => {
    if (sessions && sessions.length && firstDayOfTheWeek) {
      if (firstDayOfTheWeek === 'Monday') {
        dayjs.locale('en-gb')
      } else {
        dayjs.locale('en')
      }

      const labelsData = labels.map((label) => {
        return {
          ...label,
          sessions: 0,
          time: 0,
        }
      })

      labelsData.push({
        id: null,
        color: 'grey',
        name: 'Unlabeled',
        sessions: 0,
        time: 0,
      })

      const filterFn = getFilterFn(filter)

      sessions.forEach(({ createdAt, duration, label }) => {
        if (filterFn(createdAt, dayjs)) {
          const index = labelsData.findIndex((l) => l.id === label)

          if (index < 0) return

          labelsData[index] = {
            ...labelsData[index],
            sessions: labelsData[index].sessions + 1,
            time: labelsData[index].time + getSeconds(duration),
          }
        }
      })

      setCalculatedData(labelsData)

      setChartData((prev) => {
        const newData = { ...prev }
        newData.datasets[0].data = labelsData.map((d) => d[dataType])
        newData.datasets[0].backgroundColor = labelsData.map((l) =>
          darkModeCached
            ? materialColors[l.color][200]
            : materialColors[l.color][500]
        )
        newData.labels = labelsData.map((l) => l.name)

        return newData
      })
    }
  }, [sessions, firstDayOfTheWeek, filter, dataType, labels, darkModeCached])

  const onChipClicked = (typeSelected) => {
    if (typeSelected === dataType) return

    setDataType(typeSelected)

    setChartData((prev) => {
      const newData = { ...prev }
      newData.datasets[0].data = calculatedData.map(
        (data) => data[typeSelected]
      )
      return newData
    })

    setChartOptions((prev) => {
      const newOptions = { ...prev }

      newOptions.tooltips.callbacks.label =
        typeSelected === 'time'
          ? generateChartTimeLabel
          : generateChartSessionsLabel

      return newOptions
    })
  }

  const openMenu = (event) => {
    setAnchorEl(event.currentTarget)
  }

  const closeMenu = () => {
    setAnchorEl(null)
  }

  const handleSelect = (filter) => {
    setFilter(filter)
    closeMenu()
  }
  const getSeconds = ({ minutes, seconds }) => minutes * 60 + seconds

  return (
    <Card theme={theme} dark={darkMode}>
      <CardHeader
        title="Labels"
        action={
          <IconButton aria-label="filtering options" onClick={openMenu}>
            <MoreVertIcon />
          </IconButton>
        }
      ></CardHeader>

      <Menu
        id="long-menu"
        anchorEl={anchorEl}
        open={!!anchorEl}
        onClose={closeMenu}
        margin="dense"
      >
        {filters.map((f, i) => (
          <MenuItem
            selected={filter.value === f.value}
            key={i}
            onClick={() => handleSelect(f)}
          >
            {f.name}
          </MenuItem>
        ))}
      </Menu>

      <CardContent>
        <Typography color="textSecondary">
          {filter.displayName || filter.name || ''}
        </Typography>

        <ChartContainer>
          <Doughnut data={chartData} options={chartOptions} />
        </ChartContainer>
      </CardContent>

      <CardActions>
        <div>
          <Box mr={1} display="inline-block">
            <Chip
              onClick={() => onChipClicked('time')}
              label="Time"
              color={dataType === 'time' ? 'primary' : 'default'}
              clickable
            />
          </Box>

          <Chip
            onClick={() => onChipClicked('sessions')}
            label="Sessions"
            color={dataType === 'sessions' ? 'primary' : 'default'}
            clickable
          />
        </div>
      </CardActions>
    </Card>
  )
}

const Card = styled(MatCard)`
  background: ${({ dark }) => (dark ? '#252525' : '#fff')};
`

const CardContent = styled(MatCardContent)`
  padding-top: 0;

  .MuiTypography-colorTextSecondary {
    font-weight: bold;
    font-size: 0.9rem;
  }
`

const ChartContainer = styled.div`
  height: 250px;
  max-width: 500px;
  margin: auto;
`
