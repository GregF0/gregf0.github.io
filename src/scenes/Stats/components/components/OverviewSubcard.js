import React, { useState, useEffect } from 'react'
import { useSelector } from 'react-redux'
import styled, { css } from 'styled-components'
import {
  CardContent as MatCardContent,
  Grid,
  Box,
  Card as MatCard,
  useMediaQuery,
  useTheme,
} from '@material-ui/core'
import { TrendIcon } from './components/TrendIcon'

export const OverviewSubcard = ({ icon, sum, label, trend = 0 }) => {
  const theme = useTheme()
  const isMediumScreen = +useMediaQuery(theme.breakpoints.up('md'))
  const darkMode = useSelector((state) => +state.settings.darkMode)

  const [trendColor, setTrendColor] = useState('rgba(0,0,0,0)')
  const [trendIconPath, setTrendIconPath] = useState()

  useEffect(() => {
    if (trend === 0) return
    let color, path

    if (trend > 0) {
      if (darkMode) {
        color = trendProps.positive.dark
      } else {
        color = trendProps.positive.light
      }
      path = trendProps.positive.path
    } else {
      if (darkMode) {
        color = trendProps.negative.dark
      } else {
        color = trendProps.negative.light
      }
      path = trendProps.negative.path
    }

    setTrendColor(color)
    setTrendIconPath(path)
  }, [darkMode, trend])

  return (
    <DataCard md={isMediumScreen} dark={darkMode}>
      <CardContent md={isMediumScreen}>
        <MainGrid
          container
          direction={isMediumScreen ? 'column' : 'row'}
          spacing={2}
          alignItems="center"
        >
          <Grid item xs={2} md={5} lg={9}>
            <Icon
              max-width="100%"
              src={icon}
              alt="Today Icon"
              md={isMediumScreen}
            />
          </Grid>
          <Grid item xs={10} md={12}>
            <Sum dark={darkMode} md={isMediumScreen}>
              {sum}
            </Sum>
            <Label color={theme.palette.text.secondary} md={isMediumScreen}>
              {label}
            </Label>

            {trend !== 0 && (
              <Box
                display="flex"
                alignItems="center"
                justifyContent={isMediumScreen ? 'center' : 'flex-start'}
              >
                <TrendIcon fill={trendColor} path={trendIconPath} />
                <Trend md={isMediumScreen} color={trendColor}>
                  {Math.abs(trend)}%
                </Trend>
              </Box>
            )}
          </Grid>
        </MainGrid>
      </CardContent>
    </DataCard>
  )
}

const DataCard = styled(MatCard)`
  box-shadow: 0px 3px 3px -2px rgba(0, 0, 0, 0.2),
    0px 3px 4px 0px rgba(0, 0, 0, 0.14), 0px 1px 8px 0px rgba(0, 0, 0, 0.12);
  background: inherit;

  ${({ md }) =>
    md
      ? css`
          width: 200px;
          height: 200px;
          margin-bottom: 20px;
          background: ${({ dark }) => (dark ? '#2c2c2c' : 'inherit')};
        `
      : css`
          box-shadow: none;
          margin: 0 16px;
        `}
`

const CardContent = styled(MatCardContent)`
  ${({ md }) =>
    !md &&
    css`
      padding: 8px 0 !important;
    `}
`

const MainGrid = styled(Grid)``

const Icon = styled.img`
  width: 100%;
  max-width: ${({ md }) => (md ? '50px' : '40px')};
`

const Sum = styled.span`
  display: block;
  font-weight: bold;
  font-size: 1rem;
  color: ${({ dark }) => (dark ? '#e4e4e4' : '#4e4e4e')};

  ${({ md }) =>
    md &&
    css`
      font-size: 2rem;
      text-align: center;
    `}
`

const Label = styled.span`
  font-size: 0.8rem;
  color: ${({ color }) => color};
  display: block;
  margin-bottom: 5px;

  ${({ md }) =>
    md &&
    css`
      text-align: center;
      font-size: 0.9rem;
    `}
`

const Trend = styled.span`
  display: block;
  font-weight: bold;
  font-size: 0.9rem;
  color: ${({ color }) => color};
  margin-left: 5px;

  ${({ md }) =>
    md &&
    css`
      margin-left: 7px;
      font-size: 1rem;
    `}
`

const trendProps = {
  positive: {
    light: '#3BC83C',
    dark: '#A5D6A7',
    path: 'M6 0.806397L12 6.8064L0 6.8064L6 0.806397Z',
  },
  negative: {
    light: '#F12524',
    dark: '#EF9A9A',
    path: 'M6 6.8064L0 0.806396L12 0.806399L6 6.8064Z',
  },
}

export { Sum }
