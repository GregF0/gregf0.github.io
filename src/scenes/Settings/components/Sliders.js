import React from 'react'
import { useSelector } from 'react-redux'
import { Slider } from './components/Slider'
import {
  startSetWorkDuration,
  startSetShortBreakDuration,
  startSetLongBreakDuration,
  startSetRounds,
} from '../../../data/settings/actions'

export const Sliders = () => {
  const {
    workDuration,
    shortBreakDuration,
    longBreakDuration,
    rounds,
  } = useSelector((state) => state.settings)

  return (
    <>
      <Slider
        name="Work duration"
        marks={workMarks}
        step={5}
        min={5}
        max={60}
        unit="minutes"
        action={startSetWorkDuration}
        value={workDuration}
      />

      <Slider
        name="Short break duration"
        value={shortBreakDuration}
        marks={shortBreakMarks}
        step={5}
        min={5}
        max={30}
        unit="minutes"
        action={startSetShortBreakDuration}
      />

      <Slider
        name="Long break duration"
        value={longBreakDuration}
        marks={longBreakMarks}
        step={5}
        min={5}
        max={45}
        unit="minutes"
        action={startSetLongBreakDuration}
      />

      <Slider
        name="Rounds"
        value={rounds}
        marks={roundsMarks}
        step={1}
        min={2}
        max={15}
        unit="rounds"
        action={startSetRounds}
      />
    </>
  )
}

const workMarks = [
  {
    value: 5,
    label: '5 min',
  },
  {
    value: 25,
    label: '25 min',
  },
  {
    value: 60,
    label: '60 min',
  },
]

const shortBreakMarks = [
  {
    value: 5,
    label: '5 min',
  },
  {
    value: 30,
    label: '30 min',
  },
]

const longBreakMarks = [
  {
    value: 5,
    label: '5 min',
  },
  {
    value: 20,
    label: '20 min',
  },
  {
    value: 45,
    label: '45 min',
  },
]

const roundsMarks = [
  {
    value: 2,
    label: '2',
  },
  {
    value: 4,
    label: '4',
  },
  {
    value: 15,
    label: '15',
  },
]
