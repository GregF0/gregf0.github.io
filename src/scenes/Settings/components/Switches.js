import React, { useEffect } from 'react'
import { useSelector } from 'react-redux'
import { Switch } from './components/Switch'
import {
  startSetShowTimerInTitle,
  startSetShowNotifications,
  startSetDarkMode,
  startSetAutostart,
} from '../../../data/settings/actions'

export const Switches = () => {
  const {
    showTimerInTitle,
    showNotifications,
    darkMode,
    autostart,
  } = useSelector((state) => state.settings)

  useEffect(() => {
    if (!showTimerInTitle) {
      document.title = 'Pomodor | Productivity Timer'
    }
  }, [showTimerInTitle])

  return (
    <div>
      <Switch
        name="Timer in title"
        action={startSetShowTimerInTitle}
        checked={showTimerInTitle}
      />
      <Switch
        name="Notifications"
        action={startSetShowNotifications}
        checked={showNotifications}
      />
      <Switch name="Autostart" action={startSetAutostart} checked={autostart} />
      <Switch name="Dark mode" action={startSetDarkMode} checked={darkMode} />
    </div>
  )
}
