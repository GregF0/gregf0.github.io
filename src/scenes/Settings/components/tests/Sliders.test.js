import React from 'react'
import * as redux from 'react-redux'
import { createShallow } from '@material-ui/core/test-utils'
import { Sliders } from '../Sliders'
import { Slider } from '../components/Slider'

describe('<Sliders />', () => {
  const shallow = createShallow()
  const createWrapper = () => {
    return shallow(<Sliders />)
  }

  const createStore = () => {
    const store = {
      settings: {
        workDuration: 25,
        shortBreakDuration: 5,
        longBreakDuration: 20,
        rounds: 4,
      },
    }

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((callback) => callback(store))

    return store
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render <Sliders /> correctly', () => {
    createStore()
    expect(createWrapper()).toMatchSnapshot()
  })

  test('should render 4 sliders', () => {
    createStore()
    expect(createWrapper().find(Slider).length).toBe(4)
  })
})
