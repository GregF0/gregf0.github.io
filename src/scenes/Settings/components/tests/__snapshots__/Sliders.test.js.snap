// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sliders /> should render <Sliders /> correctly 1`] = `
<Fragment>
  <Slider
    action={[Function]}
    marks={
      Array [
        Object {
          "label": "5 min",
          "value": 5,
        },
        Object {
          "label": "25 min",
          "value": 25,
        },
        Object {
          "label": "60 min",
          "value": 60,
        },
      ]
    }
    max={60}
    min={5}
    name="Work duration"
    step={5}
    unit="minutes"
    value={25}
  />
  <Slider
    action={[Function]}
    marks={
      Array [
        Object {
          "label": "5 min",
          "value": 5,
        },
        Object {
          "label": "30 min",
          "value": 30,
        },
      ]
    }
    max={30}
    min={5}
    name="Short break duration"
    step={5}
    unit="minutes"
    value={5}
  />
  <Slider
    action={[Function]}
    marks={
      Array [
        Object {
          "label": "5 min",
          "value": 5,
        },
        Object {
          "label": "20 min",
          "value": 20,
        },
        Object {
          "label": "45 min",
          "value": 45,
        },
      ]
    }
    max={45}
    min={5}
    name="Long break duration"
    step={5}
    unit="minutes"
    value={20}
  />
  <Slider
    action={[Function]}
    marks={
      Array [
        Object {
          "label": "2",
          "value": 2,
        },
        Object {
          "label": "4",
          "value": 4,
        },
        Object {
          "label": "15",
          "value": 15,
        },
      ]
    }
    max={15}
    min={2}
    name="Rounds"
    step={1}
    unit="rounds"
    value={4}
  />
</Fragment>
`;
